source/menu/menu_coop.qc

```diff
@@ -0,0 +1,199 @@
+	// TODO: figure out why this returns 0 servers
+	// sethostcachemaskstring(0, gethostcacheindexforkey("gamedir"), cvar_string("game"), SLIST_TEST_EQUAL);
+	sethostcachesort(gethostcacheindexforkey("ping"), false);
+	refreshhostcache(false);
+	resorthostcache();
+
+	num_cached_servers = (int)gethostcachevalue(SLIST_HOSTCACHEVIEWCOUNT);
+};
+
+void() Menu_Coop_Browse =
+{
+
+	Menu_Coop_Browse_Refresh();
+
+	Menu_DrawBackground();
+
+	Menu_DrawTitle("BROWSE SERVERS");
+
+	// draw server list
+	sui_set_align([SUI_ALIGN_CENTER, SUI_ALIGN_CENTER]);
+	static vector serverlist_scrollofs;
+	vector vsize = sui_current_frame_size();
+	Menu_ServerList("cbm_serverlist", [0, 0], [vsize.x, vsize.y / 2], serverlist_scrollofs, num_cached_servers);
+
+	Menu_Button(-1, "cbm_back", "BACK", "Return to Join Game Menu.") ? current_menu = MENU_COOPJOIN : 0;
+
+	sui_pop_frame();
+};
+
+void() Menu_Coop_Direct =
+{
+	Menu_Coop_Password_Update();
+
+	Menu_DrawBackground();
+
+	Menu_DrawTitle("DIRECT IP/ID");
+
+	Menu_Button(1, "cdm_serverip", "SERVER IP/ROOM ID", "IP or ID for the Server (typically starts with /).");
+	Menu_TextInput("cdm_serverip", 1, server_ip, server_ip_cursor);
+
+	Menu_Button(2, "cdm_serverpassword", "SERVER PASSWORD", "Password for the Match set by the Host.");
+	Menu_PasswordInput("cdm_serverpassword", 2, server_password, server_password_cursor);
+
+	Menu_DrawDivider(3);
+	Menu_Button(3.25, "cdm_connect", "CONNECT TO SERVER", "Attempt to connect to Game.") ? localcmd("connect ", server_ip, "\n") : 0;
+
+	Menu_Button(-1, "cdm_back", "BACK", "Return to Join Game Menu.") ? current_menu = MENU_COOPJOIN : 0;
+
+	sui_pop_frame();
+};
+
+void() Menu_Coop_Join =
+{
+	Menu_DrawBackground();
+
+	Menu_DrawTitle("JOIN GAME");
+
+	Menu_Button(1, "cjm_browse", "BROWSE SERVERS", "Browse the list of active Servers.") ? current_menu = MENU_COOPBROWSE : 0;
+	Menu_Button(2, "cjm_direct", "DIRECT IP/ID", "Connect directly to an active Server.") ? current_menu = MENU_COOPDIRECT : 0;
+
+	Menu_Button(-1, "cjm_back", "BACK", "Return to Co-Op Menu.") ? current_menu = MENU_COOP : 0;
+
+	sui_pop_frame();
+};
+
+void() Menu_Coop_Create =
+{
+	Menu_Coop_Hostname_Update();
+	Menu_Coop_Port_Update();
+	Menu_Coop_Password_Update();
+
+	Menu_DrawBackground();
+
+	Menu_DrawTitle("CREATE GAME");
+
+	Menu_Button(1, "ccm_servername", "SERVER NAME", "Enter a name that will appear in the Server List.");
+	Menu_TextInput("ccm_servername", 1, server_hostname, server_hostname_cursor);
+
+	Menu_Button(2, "ccm_port", "SERVER PORT", "Edit the port that the Server will use.");
+	Menu_TextInput("ccm_port", 2, server_port, server_port_cursor);
+
+	Menu_Button(3, "ccm_password", "SERVER PASSWORD", "Enter an (optional) password to limit who can join.");
+	Menu_PasswordInput("ccm_password", 3, server_password, server_password_cursor);
+
+	Menu_DrawDivider(4);
+	Menu_Button(4.25, "ccm_choosemap", "CHOOSE MAP", "Select a Map to start the Game.") ? current_menu = MENU_COOPSTOCK : 0;
```

Comment:
We should redirect to a function here instead of directly setting `current_menu` that way we can set `sv_public 2`. Same applies for when the solo button is pressed, that should set `sv_public 0`.

Reply:
Resolved in 43d09e46ada4ac047b00b53f6c584ac9184e3358